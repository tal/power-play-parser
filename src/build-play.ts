import { BasicTeam, GameLiveDataPlay } from './game-data'

const startPlay: GameLiveDataPlay = {
  result: {
    event: 'Period Start',
    eventCode: 'BUF151',
    eventTypeId: 'PERIOD_START',
    description: 'Period Start',
  },
  about: {
    eventIdx: 2,
    eventId: 151,
    period: 1,
    periodType: 'REGULAR',
    ordinalNum: '1st',
    periodTime: '00:00',
    periodTimeRemaining: '20:00',
    dateTime: '2022-02-19T18:10:07Z',
    goals: { away: 0, home: 0 },
  },
  coordinates: {},
}

const endPlay: GameLiveDataPlay = {
  result: {
    event: 'Game End',
    eventCode: 'BUF675',
    eventTypeId: 'GAME_END',
    description: 'Game End',
  },
  about: {
    eventIdx: 235,
    eventId: 675,
    period: 3,
    periodType: 'REGULAR',
    ordinalNum: '3rd',
    periodTime: '20:00',
    periodTimeRemaining: '00:00',
    dateTime: '2022-02-19T20:30:48Z',
    goals: { away: 5, home: 3 },
  },
  coordinates: {},
}

const __periodEnd: GameLiveDataPlay = {
  result: {
    event: 'Period End',
    eventCode: 'BUF121',
    eventTypeId: 'PERIOD_END',
    description: 'End of 1st Period',
  },
  about: {
    eventIdx: 69,
    eventId: 121,
    period: 1,
    periodType: 'REGULAR',
    ordinalNum: '1st',
    periodTime: '20:00',
    periodTimeRemaining: '00:00',
    dateTime: '2022-02-19T18:43:54Z',
    goals: { away: 3, home: 2 },
  },
  coordinates: {},
}

const __periodStart: GameLiveDataPlay = {
  result: {
    event: 'Period Start',
    eventCode: 'BUF151',
    eventTypeId: 'PERIOD_START',
    description: 'Period Start',
  },
  about: {
    eventIdx: 2,
    eventId: 151,
    period: 1,
    periodType: 'REGULAR',
    ordinalNum: '1st',
    periodTime: '00:00',
    periodTimeRemaining: '20:00',
    dateTime: '2022-02-19T18:10:07Z',
    goals: { away: 0, home: 0 },
  },
  coordinates: {},
}

const __goal: GameLiveDataPlay = {
  players: [
    {
      player: {
        id: 8479420,
        fullName: 'Tage Thompson',
        link: '/api/v1/people/8479420',
      },
      playerType: 'Scorer',
      seasonTotal: 17,
    },
    {
      player: {
        id: 8475784,
        fullName: 'Jeff Skinner',
        link: '/api/v1/people/8475784',
      },
      playerType: 'Assist',
      seasonTotal: 15,
    },
    {
      player: {
        id: 8477949,
        fullName: 'Alex Tuch',
        link: '/api/v1/people/8477949',
      },
      playerType: 'Assist',
      seasonTotal: 13,
    },
    {
      player: {
        id: 8475311,
        fullName: 'Darcy Kuemper',
        link: '/api/v1/people/8475311',
      },
      playerType: 'Goalie',
    },
  ],
  result: {
    event: 'Goal',
    eventCode: 'BUF254',
    eventTypeId: 'GOAL',
    description:
      'Tage Thompson (17) Wrist Shot, assists: Jeff Skinner (15), Alex Tuch (13)',
    secondaryType: 'Wrist Shot',
    strength: {
      code: 'EVEN',
      name: 'Even',
    },
    gameWinningGoal: false,
    emptyNet: false,
  },
  about: {
    eventIdx: 11,
    eventId: 254,
    period: 1,
    periodType: 'REGULAR',
    ordinalNum: '1st',
    periodTime: '02:50',
    periodTimeRemaining: '17:10',
    dateTime: '2022-02-19T18:13:19Z',
    goals: {
      away: 0,
      home: 1,
    },
  },
  coordinates: {
    x: 64,
    y: 13,
  },
  team: {
    id: 7,
    name: 'Buffalo Sabres',
    link: '/api/v1/teams/7',
    triCode: 'BUF',
  },
}

const penalty: GameLiveDataPlay = {
  players: [
    {
      player: {
        id: 8481522,
        fullName: 'Peyton Krebs',
        link: '/api/v1/people/8481522',
      },
      playerType: 'PenaltyOn',
    },
    {
      player: {
        id: 8477456,
        fullName: 'J.T. Compher',
        link: '/api/v1/people/8477456',
      },
      playerType: 'DrewBy',
    },
  ],
  result: {
    event: 'Penalty',
    eventCode: 'BUF480',
    eventTypeId: 'PENALTY',
    description: 'Peyton Krebs Slashing against J.T. Compher',
    secondaryType: 'Slashing',
    penaltySeverity: 'Minor',
    penaltyMinutes: 2,
  },
  about: {
    eventIdx: 62,
    eventId: 480,
    period: 1,
    periodType: 'REGULAR',
    ordinalNum: '1st',
    periodTime: '18:44',
    periodTimeRemaining: '01:16',
    dateTime: '2022-02-19T18:39:39Z',
    goals: { away: 3, home: 2 },
  },
  coordinates: { x: 85.0, y: -21.0 },
  team: {
    id: 7,
    name: 'Buffalo Sabres',
    link: '/api/v1/teams/7',
    triCode: 'BUF',
  },
}

const teams = {
  home: {
    id: 7,
    name: 'Buffalo Sabres',
    link: '/api/v1/teams/7',
    triCode: 'BUF',
  } as BasicTeam,
  away: {
    id: 21,
    name: 'Colorado Avalanche',
    link: '/api/v1/teams/21',
    triCode: 'COL',
  } as BasicTeam,
}

export const teamIds = [teams.home.id, teams.away.id] as [number, number]

let indexTracker = 0
export function resetIndex() {
  indexTracker = 0
}

export function buildGoalPlay({
  period,
  periodTimeRemaining,
  homeAway,
}: {
  period: number
  periodTimeRemaining: string
  homeAway: 'home' | 'away'
}): GameLiveDataPlay {
  const team = teams[homeAway]

  return {
    ...__goal,
    about: {
      ...__goal.about,
      period,
      periodTimeRemaining,
      eventIdx: indexTracker++,
      eventId: indexTracker,
    },
    team,
  }
}

export function buildPenaltyPlay({
  period,
  periodTimeRemaining,
  homeAway,
  penaltyMinutes,
}: {
  period: number
  periodTimeRemaining: string
  homeAway: 'home' | 'away'
  penaltyMinutes: number
}): GameLiveDataPlay {
  const team = teams[homeAway]

  return {
    ...penalty,
    about: {
      ...penalty.about,
      period,
      periodTimeRemaining,
      eventIdx: indexTracker++,
      eventId: indexTracker,
    },
    result: {
      ...penalty.result,
      penaltyMinutes,
    },
    team,
  }
}

export function buildPeriodEndPlay({
  period,
}: {
  period: number
}): GameLiveDataPlay {
  return {
    ...__periodEnd,
    about: {
      ...__periodEnd.about,
      period,
      eventIdx: indexTracker++,
      eventId: indexTracker,
    },
  }
}

export function buildPeriodStartPlay({
  period,
}: {
  period: number
}): GameLiveDataPlay {
  return {
    ...__periodStart,
    about: {
      ...__periodStart.about,
      period,
      eventIdx: indexTracker++,
      eventId: indexTracker,
    },
  }
}

export function buildGameStartPlay(): GameLiveDataPlay {
  return {
    ...startPlay,
    about: {
      ...startPlay.about,
      eventIdx: indexTracker++,
      eventId: indexTracker,
    },
  }
}

export function buildGameEndPlay(): GameLiveDataPlay {
  return {
    ...endPlay,
    about: {
      ...endPlay.about,
      eventIdx: indexTracker++,
      eventId: indexTracker,
    },
  }
}

export type PlayTemplate = `${number}:${number} ${string}` | 'per' | 's' | 'e'
export function buildPlays(playTemplates: PlayTemplate[]): GameLiveDataPlay[] {
  let currentPeriod = 1
  const plays = [] as GameLiveDataPlay[]

  for (let t of playTemplates) {
    const [_, time, remainder] = t.match(/(?:(\d{2}:\d{2}) )?(.+)/)!
    const [type, ...args] = remainder.split(' ')
    const homeAway =
      args[0] === 'h' ? 'home' : args[0] === 'a' ? 'away' : undefined

    switch (type) {
      case 'per':
        plays.push(buildPeriodEndPlay({ period: currentPeriod }))
        currentPeriod += 1
        plays.push(buildPeriodStartPlay({ period: currentPeriod }))
        break
      case 's':
        plays.push(buildGameStartPlay())
        break
      case 'e':
        plays.push(buildGameEndPlay())
        break
      case 'g':
        if (!homeAway) throw `home or away not defined ${args[0]}`
        const goal = buildGoalPlay({
          homeAway,
          period: currentPeriod,
          periodTimeRemaining: time,
        })
        plays.push(goal)
        break
      case 'p':
        if (!homeAway) throw `home or away not defined ${args[0]}`
        const penaltyMinutes = parseInt(args[1], 10)
        if (isNaN(penaltyMinutes))
          throw `value isn't number for penalty length ${penaltyMinutes}`
        const penalty = buildPenaltyPlay({
          homeAway,
          period: currentPeriod,
          periodTimeRemaining: time,
          penaltyMinutes,
        })
        plays.push(penalty)
        break
      default:
        throw `unparsable template ${t}`
    }
  }

  return plays
}
