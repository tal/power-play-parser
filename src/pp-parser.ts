import { Double } from './double'
import { BasicTeam, GameLiveDataPlay } from './game-data'
import { PowerPlay } from './power-play'
import { getTimeIntoGame } from './time-into-game'

export type ParsePowerPlaysParams = {
  teamIds?: [number, number]
}

export function parsePowerPlays(
  plays: GameLiveDataPlay[],
  params: ParsePowerPlaysParams = {}
) {
  const powerPlays: Record<number, PowerPlay> = {}

  let teams = new Double<BasicTeam['id']>(params.teamIds)
  for (let play of plays) {
    const teamId = play.team?.id
    if (teamId) {
      teams.add(teamId)
    }

    if (teams.isComplete) {
      break
    }
  }

  const [team1Id, team2Id] = teams.values

  if (!team1Id || !team2Id) {
    throw `cannot parse plays without both teams`
  }

  const activePowerPlays = {
    [team1Id]: [] as PowerPlay[],
    [team2Id]: [] as PowerPlay[],
  }

  function removePowerPlaysThatHaveExpired(by: number) {
    function filterOutExpiredPPs(pp: PowerPlay): boolean {
      if (pp.isAfterExpectedDuration(by)) {
        pp.actualDurationInSeconds = pp.expectedDurationSeconds
        powerPlays[pp.playIndex] = pp
        return false
      } else {
        return true
      }
    }

    activePowerPlays[team1Id!] =
      activePowerPlays[team1Id!].filter(filterOutExpiredPPs)
    activePowerPlays[team2Id!] =
      activePowerPlays[team2Id!].filter(filterOutExpiredPPs)
  }

  let lastPlayStartedAt = -1
  for (let play of plays) {
    const playTimeSeconds = getTimeIntoGame(play.about)
    if (playTimeSeconds < lastPlayStartedAt) {
      throw `All plays must be in order, idx:${play.about.eventIdx} was at ${playTimeSeconds}, previous was at ${lastPlayStartedAt}`
    }
    lastPlayStartedAt = playTimeSeconds
    removePowerPlaysThatHaveExpired(playTimeSeconds)

    if (play.result.eventTypeId === 'GAME_END') {
      activePowerPlays[team1Id!].forEach((pp) => {
        pp.actualEndSeconds = playTimeSeconds
      })
      activePowerPlays[team2Id!].forEach((pp) => {
        pp.actualEndSeconds = playTimeSeconds
      })
    }

    const team = play.team
    if (!team) {
      continue
    }
    const teamId = team.id
    const otherTeamId = teamId ? teams.other(teamId) : undefined

    function teamsFirstPPEnds(teamIdToEnd: typeof teamId | undefined) {
      if (!teamIdToEnd) return

      const pps = activePowerPlays[teamIdToEnd] //??
      const ppToEnd = pps.shift()
      if (!ppToEnd) return

      const timeIntoPP = playTimeSeconds - ppToEnd.startSeconds

      if (ppToEnd.expectedDurationSeconds === 60 * 2) {
        ppToEnd.actualDurationInSeconds = timeIntoPP
        return ppToEnd
      } else if (ppToEnd.expectedDurationSeconds === 60 * 4) {
        // less than 2 min in
        if (timeIntoPP < 120) {
          ppToEnd.expectedDurationSeconds -= 120 - timeIntoPP
          pps.unshift(ppToEnd)
          return
        } else {
          ppToEnd.actualDurationInSeconds = timeIntoPP
          return ppToEnd
        }
      } else if (ppToEnd.expectedDurationSeconds === 60 * 5) {
        pps.unshift(ppToEnd)
        return
      }
    }

    if (play.result.eventTypeId === 'PENALTY') {
      const otherTeamsEndedPP = teamsFirstPPEnds(otherTeamId)

      let pp = new PowerPlay(
        play.about.eventIdx,
        teamId,
        playTimeSeconds,
        play.result.penaltyMinutes! * 60
      )

      if (otherTeamsEndedPP) {
        pp.expectedDurationSeconds = otherTeamsEndedPP.actualDurationInSeconds!
        const timeDiff =
          otherTeamsEndedPP.expectedDurationSeconds -
          otherTeamsEndedPP.actualDurationInSeconds!
        pp.startSeconds += timeDiff
      }

      activePowerPlays[teamId].push(pp)
      powerPlays[pp.playIndex] = pp
    }

    if (play.result.eventTypeId === 'GOAL') {
      teamsFirstPPEnds(otherTeamId)
    }
  }

  // Remove any power plays that have zero duration or were never actually set
  const filtredPPs = Object.values(powerPlays).filter(
    (pp) => pp.actualDurationInSeconds ?? 0 > 0
  )

  return { powerPlays: filtredPPs }
}
