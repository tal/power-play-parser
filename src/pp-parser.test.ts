import { buildPlays, PlayTemplate, teamIds } from './build-play'
import { parsePowerPlays as ogParsePowerPlays } from './pp-parser'

function parsePowerPlays(templates: PlayTemplate[]) {
  const plays = buildPlays(['s', ...templates, 'e'])
  return ogParsePowerPlays(plays, { teamIds: teamIds })
}

type Parse = ReturnType<typeof parsePowerPlays>
type PP = Parse['powerPlays'][number]

describe('parsePowerPlays', () => {
  it('should return empty', () => {
    const plays: any[] = []
    const pp = parsePowerPlays(plays)
    expect(pp.powerPlays.length).toBe(0)
  })

  it('should have one complete pp', () => {
    const pp = parsePowerPlays(['10:00 p a 2', '05:00 g a'])
    expect(pp.powerPlays.length).toBe(1)
    expect(pp.powerPlays[0].actualDurationInSeconds).toBe(120)
  })

  it('should have two complete pps', () => {
    const pp = parsePowerPlays([
      '10:00 p a 2',
      '05:00 g a',
      'per',
      '10:00 p a 2',
    ])
    expect(pp.powerPlays.length).toBe(2)
    expect(pp.powerPlays[0].actualDurationInSeconds).toBe(120)
  })

  it('should have one partial pp', () => {
    const pp = parsePowerPlays(['10:00 p h 2', '09:00 g a'])
    expect(pp.powerPlays.length).toBe(1)
    expect(pp.powerPlays[0].actualDurationInSeconds).toBe(60)
  })

  it('is one 5 on 3', () => {
    const pp = parsePowerPlays(['10:10 p h 2', '10:00 p h 2'])
    expect(pp.powerPlays.length).toBe(2)
    expect(pp.powerPlays[0].actualDurationInSeconds).toBe(120)
    expect(pp.powerPlays[1].actualDurationInSeconds).toBe(120)
  })

  it('should error if out of order', () => {
    expect(() =>
      parsePowerPlays(['10:00 p h 2', '10:10 p h 2'])
    ).toThrowErrorMatchingSnapshot()
  })

  it('shortened pp', () => {
    const pp = parsePowerPlays(['10:10 p h 2', '10:00 p a 2'])
    expect(pp.powerPlays.length).toBe(2)
    const pp1 = pp.powerPlays[0]
    const pp2 = pp.powerPlays[1]
    expect(pp1.actualDurationInSeconds).toBe(10)
    expect(pp2.actualDurationInSeconds).toBe(10)
    expect(pp2.startSeconds).toBe(pp1.actualEndSeconds! + 120 - 10)
  })

  it('3v5 with goal', () => {
    const pp = parsePowerPlays(['10:20 p a 2', '10:10 p a 2', '10:00 g h'])
    expect(pp.powerPlays.length).toBe(2)
    const pp1 = pp.powerPlays[0]
    const pp2 = pp.powerPlays[1]
    expect(pp1.actualDurationInSeconds).toBe(20)
    expect(pp2.actualDurationInSeconds).toBe(120)
    expect(pp2.startSeconds).toEqual(600 - 10)
  })

  describe('two offsetting pps', () => {
    let pp: Parse
    beforeEach(() => {
      pp = parsePowerPlays(['10:00 p a 2', '10:00 p h 2'])
    })

    it('should have correct count', () => {
      expect(pp.powerPlays.length).toBe(0)
    })
  })

  describe('two offsetting pps and a goal', () => {
    let pp: Parse
    beforeEach(() => {
      pp = parsePowerPlays(['10:00 p a 2', '10:00 p h 2', '09:30 g h'])
    })

    it('should have correct count', () => {
      expect(pp.powerPlays.length).toBe(0)
    })
  })

  it('4min', () => {
    const pp = parsePowerPlays(['10:10 p a 4'])
    expect(pp.powerPlays.length).toBe(1)
    expect(pp.powerPlays[0].actualDurationInSeconds).toBe(60 * 4)
  })

  it('4min with early goal', () => {
    const pp = parsePowerPlays(['10:10 p a 4', '10:00 g h'])
    expect(pp.powerPlays.length).toBe(1)
    expect(pp.powerPlays[0].actualDurationInSeconds).toBe(120 + 10)
  })

  it('4min with early goal', () => {
    const pp = parsePowerPlays(['10:00 p a 4', '08:01 g h'])
    expect(pp.powerPlays.length).toBe(1)
    expect(pp.powerPlays[0].actualDurationInSeconds).toBe(240 - 1)
  })

  it('4min with late goal', () => {
    const pp = parsePowerPlays(['10:00 p a 4', '07:59 g h'])
    expect(pp.powerPlays.length).toBe(1)
    expect(pp.powerPlays[0].actualDurationInSeconds).toBe(120 + 1)
  })

  it('5min with late goals', () => {
    const pp = parsePowerPlays([
      '10:00 p a 5',
      '08:59 g h',
      '07:59 g h',
      '07:59 g h',
    ])
    expect(pp.powerPlays.length).toBe(1)
    expect(pp.powerPlays[0].actualDurationInSeconds).toBe(5 * 60)
  })

  describe('two offsetting pps then another pp', () => {
    let pp: Parse
    let pp1: PP | undefined
    let pp2: PP | undefined
    beforeEach(() => {
      pp = parsePowerPlays(['10:00 p a 2', '10:00 p h 2', '09:30 p h 2'])
      pp1 = pp.powerPlays[0]
      pp2 = pp.powerPlays[1]
    })

    it('should have correct count', () => {
      expect(pp.powerPlays.length).toBe(1)
    })

    it('should have correct duration', () => {
      expect(pp1?.actualDurationInSeconds).toBe(120)
    })
  })

  describe('power plays across boundaries', () => {
    let pp: Parse
    let pp1: PP | undefined
    let pp2: PP | undefined
    beforeEach(() => {
      pp = parsePowerPlays(['01:00 p a 2', 'per', '19:59 g h'])
      pp1 = pp.powerPlays[0]
      pp2 = pp.powerPlays[1]
    })

    it('should have correct count', () => {
      expect(pp.powerPlays.length).toBe(1)
    })

    it('should have correct duration', () => {
      expect(pp1?.actualDurationInSeconds).toBe(61)
    })
  })

  describe('4min with 2min in middle', () => {
    let pp: Parse
    let pp1: PP | undefined
    let pp2: PP | undefined
    beforeEach(() => {
      pp = parsePowerPlays(['10:00 p a 4', '09:00 p h 2'])
      pp1 = pp.powerPlays[0]
      pp2 = pp.powerPlays[1]
    })

    it('should have correct count', () => {
      expect(pp.powerPlays.length).toBe(2)
    })

    it('should have correct duration', () => {
      expect(pp1?.actualDurationInSeconds).toBe(60)
    })

    it('should have correct team', () => {
      expect(pp1?.teamId).toBe(21)
    })
    it('should have correct duration', () => {
      expect(pp2?.actualDurationInSeconds).toBe(60)
    })

    it('should have correct team', () => {
      expect(pp2?.teamId).toBe(21)
    })
  })

  describe('crazy1', () => {
    let pp: Parse
    let pp1: PP | undefined
    let pp2: PP | undefined
    beforeEach(() => {
      pp = parsePowerPlays([
        '10:00 p a 4',
        '09:00 p h 2',
        '09:00 p a 2',
        '08:31 g h',
        '08:30 g a',
      ])
      pp1 = pp.powerPlays[0]
      pp2 = pp.powerPlays[1]
    })

    it('should have correct count', () => {
      expect(pp.powerPlays.length).toBe(1)
    })
  })

  describe('end of game', () => {
    let pp: Parse
    let pp1: PP | undefined
    let pp2: PP | undefined
    beforeEach(() => {
      pp = parsePowerPlays([
        // '10:10 g a',
        // '10:00 g h',
        'per',
        'per',
        '01:00 p a 2',
      ])
      pp1 = pp.powerPlays[0]
      pp2 = pp.powerPlays[1]
    })

    it('should have correct count', () => {
      expect(pp.powerPlays.length).toBe(1)
    })

    it('should have correct duration', () => {
      expect(pp1?.actualDurationInSeconds).toBe(60)
    })
  })
})
