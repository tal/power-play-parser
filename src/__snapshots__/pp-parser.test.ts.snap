// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsePowerPlays should error if out of order 1`] = `undefined`;
